service/api/service.go

```diff
@@ -156,15 +156,30 @@ func (s *serviceImpl) ApiV1WorkflowWaitForStateCompletion(
 ) (wresp *iwfidl.WorkflowWaitForStateCompletionResponse, retError *errors.ErrorAndStatus) {
 	defer func() { log.CapturePanic(recover(), s.logger, &retError) }()
 
-	var workflowId string
+	var parentWfId, currentWorkflowId, _ string
+
+	response, err := s.client.DescribeWorkflowExecution(ctx, req.GetWorkflowId(), "", nil)
+	if err != nil {
+		return nil, s.handleError(err, WorkflowWaitForStateCompletionApiPath, req.WorkflowId)
+	}
+
+	if response.FirstRunId == "" {
+		parentWfId = req.WorkflowId // Cadence
```

Comment:
If we know cadence doesn't have this, we can avoid calling the `DescribeWorkflowExecution` API for cadence. 

Although this API is really fast (it's backed by caching on cadence/temporal) but still nice to do so if this is not too much. (we do have cadence users in OSS, some ppl just perfer using cadence)

Reply:
Added GetBackendType method to UnifiedClient